File: js/test_money.js
// 编写第1个测试用例
const assert = require('assert');
const Money = require('./money');
const Portfolio = require('./portfolio');
//将测试规整到1个类里
class MoneyTest{
    testMultiplication() {
        // 测试代码1，针对乘法
        let tenEuros = new Money(10, "EUR");
        let twentyEuros = new Money(20, "EUR");
        assert.deepStrictEqual(tenEuros.times(2), twentyEuros);

    }
    testDivision(){
        // 测试代码2，针对除法
        let originalMoney = new Money(4002, "KRW")
        let actualMoneyAfterDivision = originalMoney.divide(4)
        let expectedMoneyAfterDivision = new Money(1000.5, "KRW")
        assert.deepStrictEqual(actualMoneyAfterDivision, expectedMoneyAfterDivision)
    }
    testAddition() {
    // 测试代码3，针对汇率转换
    let fiveDollars = new Money(5, "USD");
    let tenDollars = new Money(10, "USD")
    let fifteenDollars = new Money(15, "USD")
    let portfolio = new Portfolio();
    portfolio.add(fiveDollars, tenDollars);
    assert.deepStrictEqual(portfolio.evaluate("USD"), fifteenDollars);
    }

    // 测试代码4，美元与欧元相加
    testAdditionOfDollarsAndEuros() {
        let fiveDollars = new Money(5, "USD");
        let tenEuros = new Money(10, "EUR");
        let portfolio = new Portfolio();
        portfolio.add(fiveDollars, tenEuros);
        let expectedValue = new Money(17, "USD");
        assert.deepStrictEqual(portfolio.evaluate("USD"), expectedValue);
    }

    //测试代码5，美元与韩元相加
    testAdditionOfDollarsAndWons() {
        let oneDollar = new Money(1, "USD");
        let elevenHundredWon = new Money(1100, "KRW");
        let portfolio = new Portfolio();
        portfolio.add(oneDollar, elevenHundredWon);
        let expectedValue = new Money(2200, "KRW");
        assert.deepStrictEqual(portfolio.evaluate("KRW"), expectedValue);
    }

    getAllTestMethods() {
        let moneyPrototype = MoneyTest.prototype; // 获取MoneyTest类的原型对象
        let allProps = Object.getOwnPropertyNames(moneyPrototype); // 获取原型对象的所有属性名
        let testMethods = allProps.filter(p => {
            return typeof moneyPrototype[p] === 'function' && p.startsWith('test'); // 过滤出以test开头的方法
        });
        return testMethods;
    }

    runAllTests() {
        let testMethods = this.getAllTestMethods(); // 获取所有的测试方法名
        testMethods.forEach(m => {
            // 测试触发该方法之前，先打印出其名字
            console.log("Running: %s()", m);
            let method = Reflect.get(this, m); // 获取该方法的method对象
            try {
                Reflect.apply(method, this, []); // 在this对象上调用method所表示的测试方法
            } catch (e) {
                if ( e instanceof assert.AssertionError) {
                    console.log(e);
                } else {
                    throw e;
                }
            }
            
        })
    }
}

//执行所有测试
new MoneyTest().runAllTests();

File: py/test_money.py
# 编写第一个TDD测试用例，2026年03月17日09:00:00
import unittest
import functools
import operator
from money import Money
from portfolio import Portfolio

# 测试用例代码1
class TestMoney(unittest.TestCase):
    
    # 测试代码1
    def testMultiplication(self):
        tenEuros = Money(10, "EUR")
        twentyEuros = tenEuros.times(2)
        self.assertEqual(20, twentyEuros.amount)
        self.assertEqual("EUR", twentyEuros.currency)
        
    # 测试代码2
    def testDivision(self):
        originalMoney = Money(4002, "KRW")
        actualMoneyAfterDivision = originalMoney.divide(4)
        expectedMoneyAfterDivision = Money(1000.5, "KRW")
        self.assertEqual(expectedMoneyAfterDivision.amount,
                         actualMoneyAfterDivision.amount)
        self.assertEqual(expectedMoneyAfterDivision.currency,
                         actualMoneyAfterDivision.currency)    
        
    # 测试代码3
    def testAddition(self):
        fiveDollars = Money(5, "USD")
        tenDollars = Money(10, "USD")
        fifteenDollars = Money(15, "USD")
        portfolio = Portfolio()
        portfolio.add(fiveDollars, tenDollars)
        self.assertEqual(fifteenDollars, portfolio.evaluate("USD"))
        
    # 测试代码4 测试美元与欧元相加
    def testAdditionOfDollarsAndEuros(self):
        fiveDollars = Money(5, "USD")
        tenEuros = Money(10, "EUR")
        portfolio = Portfolio()
        portfolio.add(fiveDollars, tenEuros)
        expectedValue = Money(17, "USD")
        actualValue = portfolio.evaluate("USD")
        self.assertEqual(expectedValue, actualValue, "%s != %s" % (expectedValue, actualValue))
    
    # 测试代码5 测试美元与韩元相加
    def testAdditionOfDollarsAndWons(self):
        oneDollar = Money(1, "USD")
        elevenHundredWon = Money(1100, "KRW")
        portfolio = Portfolio()
        portfolio.add(oneDollar, elevenHundredWon)
        expectedValue = Money(2200, "KRW")
        actualValue = portfolio.evaluate("KRW")
        self.assertEqual(
            expectedValue, actualValue, "%s != %s" %(expectedValue, actualValue)
        )    
if __name__ == '__main__':
    unittest.main()